spectrogram-feature-extraction.html

```diff
@@ -0,0 +1,468 @@
+    x = np.absolute(x)**2
+    x[1:-1, :] *= (2.0 / scale)
+    x[(0, -1), :] /= scale
+    freqs = float(sample_rate) / fft_length * np.arange(x.shape[0])
+    ind = np.where(freqs <= max_freq)[0][-1] + 1
+    result = np.transpose(np.log(x[:ind, :] + eps))
+
+    return result`
+}
+py(_spectrogram());
+
+%% md
+# Decode wav utilities
+
+This is utility code to decode our .wav file and get channel data and sample rate
+
+%% js
+// decode wav
+function pcm16(buffer, offset, output, channels, samples){
+  let input = new Int16Array(buffer, offset);
+  let pos = 0;
+  for (let i = 0; i < samples; ++i) {
+    for (let ch = 0; ch < channels; ++ch) {
+      let data = input[pos++];
+      output[ch][i] = data < 0 ? data / 32768 : data / 32767;
+    }
+  }
+}
+function decode(buffer) {
+  let pos = 0, end = 0;
+  if (buffer.buffer) {
+    // If we are handed a typed array or a buffer, then we have to consider the
+    // offset and length into the underlying array buffer.
+    pos = buffer.byteOffset;
+    end = buffer.length;
+    buffer = buffer.buffer;
+  } else {
+    // If we are handed a straight up array buffer, start at offset 0 and use
+    // the full length of the buffer.
+    pos = 0;
+    end = buffer.byteLength;
+  }
+
+  let v = new DataView(buffer);
+
+  function u8() {
+    let x = v.getUint8(pos);
+    pos++;
+    return x;
+  }
+
+  function u16() {
+    let x = v.getUint16(pos, true);
+    pos += 2;
+    return x;
+  }
+
+  function u32() {
+    let x = v.getUint32(pos, true);
+    pos += 4;
+    return x;
+  }
+
+  function string(len) {
+    let str = '';
+    for (let i = 0; i < len; ++i)
+      str += String.fromCharCode(u8());
+    return str;
+  }
+
+  if (string(4) !== 'RIFF')
+    throw new TypeError('Invalid WAV file');
+  u32();
+  if (string(4) !== 'WAVE')
+    throw new TypeError('Invalid WAV file');
+
+  let fmt;
+
+  while (pos < end) {
+    let type = string(4);
+    let size = u32();
+    let next = pos + size;
+    switch (type) {
+    case 'fmt ':
+      let formatId = u16();
+      if (formatId !== 0x0001 && formatId !== 0x0003)
+        throw new TypeError('Unsupported format in WAV file: ' + formatId.toString(16));
+      fmt = {
+        format: 'lpcm',
+        floatingPoint: formatId === 0x0003,
+        channels: u16(),
+        sampleRate: u32(),
+        byteRate: u32(),
+        blockSize: u16(),
+        bitDepth: u16(),
+      };
+      break;
+    case 'data':
+      if (!fmt)
+        throw new TypeError('Missing "fmt " chunk.');
+      let samples = Math.floor(size / fmt.blockSize);
+      let channels = fmt.channels;
+      let sampleRate = fmt.sampleRate;
+      let channelData = [];
+      for (let ch = 0; ch < channels; ++ch)
+        channelData[ch] = new Float32Array(samples);
+      pcm16(buffer, pos, channelData, channels, samples);
+      return {
+        sampleRate: sampleRate,
+        channelData: channelData
+      };
+      break;
+    }
+    pos = next;
+  }
+}
```

Comment:
Is there a reason not to just use Web Audio's `decodeAudioData` here?

Reply:
i have absolute no control over `decodeAudioData`, sometimes it returns me a decoded buffer with the wrong `sampleRate` for some reason, even if the `arrayBuffer`  comes from a `16000 sampleRate` recording. here i can manually use a `pcm16` decoder on the buffer